api/src/plugins/shadow-capture.ts

```diff
@@ -0,0 +1,137 @@
+import { randomUUID } from 'crypto';
+import { appendFileSync, mkdirSync } from 'fs';
+import { join } from 'path';
+import type { FastifyPluginCallback } from 'fastify';
+
+import fp from 'fastify-plugin';
+import { FastifyReply } from 'fastify/types/reply';
+import { FastifyRequest } from 'fastify/types/request';
+
+const LOGS_DIRECTORY = 'logs';
+const REQUEST_CAPTURE_FILE = 'request-capture.jsonl';
+const RESPONSE_CAPTURE_FILE = 'response-capture.jsonl';
+
+let REQUEST_BUFFER: unknown[] = [];
+let RESPONSE_BUFFER: unknown[] = [];
+
+/**
+ * Plugin for capturing requests and responses to allow shadow testing.
+ *
+ * @param fastify The Fastify instance.
+ * @param _options Options passed to the plugin via `fastify.register(plugin, options)`.
+ * @param done Callback to signal that the logic has completed.
+ */
+const shadowCapture: FastifyPluginCallback = (fastify, _options, done) => {
+  mkdirSync(LOGS_DIRECTORY, { recursive: true });
+  fastify.addHook('onRequest', (req, rep, done) => {
+    // Attach timestamp at beginning of lifecycle
+    // @ts-expect-error Exists
+    req.__timestamp = Date.now();
+
+    // Give request and response same id to match.
+    const id = randomUUID();
+    // @ts-expect-error Exists
+    req.__id = id;
+    // @ts-expect-error Exists
+    rep.__id = id;
+    done();
+  });
+
+  // Body is only included after `Parsing` lifecycle
+  fastify.addHook('preValidation', (req, rep, done) => {
+    captureRequest(req);
+    done();
+  });
+
+  fastify.addHook('onSend', async (_req, rep, payload) => {
+    // @ts-expect-error Exists
+    rep.__payload = payload;
+    return payload;
+  });
+
+  fastify.addHook('onResponse', (_req, rep, done) => {
+    captureReply(rep);
+    done();
+  });
+
+  done();
+};
+
+/* eslint-disable @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-return */
+function captureRequest(req: FastifyRequest) {
+  const savedRequest = {
+    // @ts-expect-error Exists
+    id: req.__id,
+    // @ts-expect-error Exists
+    timestamp: req.__timestamp,
+    url: req.url,
+    headers: omit(req.headers, 'cookie'),
+    cookies: include(req.cookies, '_csrf', 'csrf_token', 'jwt_access_token'),
+    user: req.user,
+    body: req.body
+  };
+
+  if (REQUEST_BUFFER.length > 10) {
+    appendFileSync(
+      join(LOGS_DIRECTORY, REQUEST_CAPTURE_FILE),
+      REQUEST_BUFFER.map(rb => JSON.stringify(rb)).join('\n') + '\n'
+    );
+    REQUEST_BUFFER = [savedRequest];
+  } else {
+    REQUEST_BUFFER.push(savedRequest);
+  }
```

Comment:
```suggestion
  appendFileSync(
    join(LOGS_DIRECTORY, REQUEST_CAPTURE_FILE),
    JSON.stringify(savedRequest) + '\n'
  );
```
Why not this?

Reply:
Because, if we want to try capture 100's of requests and look at their timings, a lack of a buffer would significantly affect the results.

_"One does not simply know both the data and performance of a request, without using a buffer."_ - Werner Heisenberg, 1927 [slightly modified]

Reply:
Sure, but this displaces the problem, rather than solving it. With this approach you get all the delay at once.

I don't know why the timing matters, exactly, but isn't that worse?